src/main/scala/lantern/TensorDifferentiation.scala

```diff
@@ -688,6 +709,35 @@ trait TensorDsl extends DslOps with Diff {
         }
       }
     }
+
+    @virtualize
+    override def relu(x: Tensor): Tensor = {
+      val res = backend.mallocArray[Float](x.scalarCount)
+      for (i <- 0 until x.scalarCount: Rep[Range]) {
+        if (x.data(i) < 0.0f)
+          res(i) = 0.0f
+        else
+          res(i) = x.data(i)
+      }
+      Tensor(res, x.shape.seq : _*)
+    }
+
+    @virtualize
+    override def relu_grad(input: TensorR, res: TensorR): Unit = {
+      for (i <- 0 until input.x.scalarCount: Rep[Range]) {
+        input.d.data(i) = if (input.x.data(i) < 0.0f) 0.0f else res.d.data(i)
+      }
+    }
+
+    override def tanh(x: Tensor) = x.map(s => Math.tanh(s).toFloat)
```

Comment:
It'd be good to decouple these backend functions from `Tensor` methods (`map` and `add_oneMinusSquare_mult`).

Reply:
yes, I agree. We are sort things out as we encounter them.